pkg/workflow/controllers/chaos_node_reconciler.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+)
+
+type ChaosNodeReconciler struct {
+	kubeClient    client.Client
+	eventRecorder record.EventRecorder
+	logger        logr.Logger
+}
+
+func NewChaosNodeReconciler(kubeClient client.Client, eventRecorder record.EventRecorder, logger logr.Logger) *ChaosNodeReconciler {
+	return &ChaosNodeReconciler{kubeClient: kubeClient, eventRecorder: eventRecorder, logger: logger}
+}
+
+func (it *ChaosNodeReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.TODO()
+	node := v1alpha1.WorkflowNode{}
+
+	err := it.kubeClient.Get(ctx, request.NamespacedName, &node)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if ConditionEqualsTo(node.Status, v1alpha1.ConditionDeadlineExceed, corev1.ConditionTrue) {
+		err := it.recoverChaos(ctx, node)
+		return reconcile.Result{}, err
+	}
+
+	if availableChaos(string(node.Spec.Type)) {
+		err = it.injectChaos(ctx, node)
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (it *ChaosNodeReconciler) injectChaos(ctx context.Context, node v1alpha1.WorkflowNode) error {
+	var chaosObject runtime.Object
+	var meta metav1.Object
+
+	if node.Spec.Type == v1alpha1.TypeNetworkChaos {
```

Comment:
rly? one if per chaos type?

Reply:
There're only several limited chaos types for the demo. 😰

It should be one method like `fillChaosSpec()` in the future.